### DISCUSSION TITLE
Vaultwarden wont import cvs file

### DISCUSSION THREAD

**Discussion by wioo** [dani-garcia/vaultwarden#3426] [Q&A] [answered]
### Subject of the issue
Vaultwarden wont import basic Bitwarden cvs file.

### Deployment environment
* Vaultwarden version: v1.28.1
* Web-vault version: v2023.3.0b
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: false (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.39.2
* Clients used: web browser
* Reverse proxy and version: nginx proxy manager  v2.10.2
* Other relevant information: works with 1.27.0

### Steps to reproduce
Didnt change anything. When vaulwarden was updated to 1.28.0 import stopped working.

### Expected behaviour
Sucessfully Imported passwords.

### Actual behaviour
Nothing happens when import button is clicked.

### Troubleshooting data
Firefox console errors when I open vaultwarden login page
```
Content Security Policy: The page’s settings blocked the loading of a resource at eval (“script-src”)
Source: HTMLElement.prototype.focus = ((realFocu….

TypeError: can't convert undefined to object
```
When try to import this error is presented
```
TypeError: can't convert undefined to object

TypeError: t is undefined                      consoleLog.service.ts:51:16
    processFolder base-importer.ts:416
    parse bitwarden-csv-importer.ts:48
    parse bitwarden-csv-importer.ts:24
    import import.service.ts:109
    mt 297.eab0517393b948250535.js:1
    t zone.js:1351
    mt 297.eab0517393b948250535.js:1
    import 297.eab0517393b948250535.js:1
    submit import.component.ts:117
    o 297.eab0517393b948250535.js:1
    invoke zone.js:409
    onInvoke ng_zone.ts:418
    invoke zone.js:408
    run zone.js:169
    C zone.js:1326
    invokeTask zone.js:443
    onInvokeTask ng_zone.ts:403
    invokeTask zone.js:442
    runTask zone.js:214
    b zone.js:632
    promise callback*y zone.js:608
    m zone.js:619
    scheduleTask zone.js:433
    onScheduleTask zone.js:320
    scheduleTask zone.js:423
    scheduleTask zone.js:257
    scheduleMicroTask zone.js:277
    C zone.js:1316
    P zone.js:1253
    O zone.js:1169
    A zone.js:1185
    onload import.component.ts:233
    invoke zone.js:409
    onInvoke ng_zone.ts:418
    invoke zone.js:408
    runGuarded zone.js:180
    s zone.js:163
    set zone.js:980
    getFileContents import.component.ts:220
    t zone.js:1351
    getFileContents import.component.ts:217
    submit import.component.ts:97
    E 297.eab0517393b948250535.js:1
    t zone.js:1351
    E 297.eab0517393b948250535.js:1
    submit 297.eab0517393b948250535.js:1
    template import.component.html:7
    jc listener.ts:225
    r listener.ts:263
    next Subscriber.ts:161
    _next Subscriber.ts:119
    next Subscriber.ts:75
    next Subject.ts:68
    r errorContext.ts:29
    next Subject.ts:61
    emit event_emitter.ts:114
    onSubmit ng_form.ts:300
    hostBindings ng_form.ts:136
    jc listener.ts:225
    r listener.ts:263
    _ dom_renderer.ts:67
    invokeTask zone.js:443
    onInvokeTask ng_zone.ts:403
    invokeTask zone.js:442
    runTask zone.js:214
```
Same with qutebrowser on linux and firefox & chrome on windows.

**Comment by BlackDex** [comment]
Looks like a client side error.
Either the csv is invalid, or something else is wrong.

What happens in an other browser?
Does the developer console show errors during the initial load during login/page load. Like not being able to access the site?

**Comment by wioo** [comment]
qutebrowser 
on login page
```
The source list for Content Security Policy directive 'script-src' contains an invalid source: ''wasm-unsafe-eval''. It will be ignored.
```
when importing 
```
TypeError: Cannot read property 'replace' of undefined
    at C.processFolder (base-importer.ts:416)
    at bitwarden-csv-importer.ts:48
    at Array.forEach (<anonymous>)
    at C.parse (bitwarden-csv-importer.ts:24)
    at ft.<anonymous> (import.service.ts:109)
    at Generator.next (<anonymous>)
    at 297.eab0517….js:1
    at new t (zone.js:1351)
    at mt (297.eab0517….js:1)
    at ft.import (297.eab0517….js:1)
write	@	consoleLog.service.ts:51
...
```

firefox on win (v 111.0.1)
no errors on login page

So just 
```
TypeError: t is undefined consoleLog.service.ts:51:16 
...
```

when importing (same as linux)

Chrome on win
on login page
```
Error with Permissions-Policy header: Unrecognized feature: 'ambient-light-sensor'.
Error with Permissions-Policy header: Unrecognized feature: 'battery'.
Error with Permissions-Policy header: Unrecognized feature: 'document-domain'.
Error with Permissions-Policy header: Unrecognized feature: 'execution-while-not-rendered'.
Error with Permissions-Policy header: Unrecognized feature: 'execution-while-out-of-viewport'.
```
when importing
```consoleLog.service.ts:51 TypeError: Cannot read properties of undefined (reading 'replace')
    at C.processFolder (base-importer.ts:416:29)
    at bitwarden-csv-importer.ts:48:14
    at Array.forEach (<anonymous>)
    at C.parse (bitwarden-csv-importer.ts:24:13)
    at ft.<anonymous> (import.service.ts:109:41)
    at Generator.next (<anonymous>)
    at 297.eab0517393b948250535.js:1:160136
    at new t (zone.js:1351:25)
    at mt (297.eab0517393b948250535.js:1:159881)
    at ft.import (297.eab0517393b948250535.js:1:160483)
...
```

Version 1.27.0
qutebrowser without errors.

firefox (linux) the same error at login page but import works.

**Reply by stefan0xC** [reply]
Can you please use \`\`\`  to format the output for better readability? (You can select the lines and press <kbd>Ctrl</kbd>+<kbd>e</kbd>)

**Reply by wioo** [reply]
Sorry. Done.

**Comment by BlackDex** [comment]
Looks like maybe some reverse proxy issue. Caching, ModSecurity, other headers which might get replaced.
Or browser extensions which block or modify something.

But if it does work in an other browser it tells me it works.
Also, it's not a server side issue, and it also not looks like a client issue. But if it is an issue, then it's not something we can fix.

**Comment by wioo** [comment]
There is a chance that reverse proxy is at fault. But why does 1.27.0 work in all browsers and 1.28.1 doesnt in neither? 
This also happens in qutebrowser --temp-basedir which doesnt contain any extensions etc.

I also downgraded reverse proxy and recreated current version but nothing changes.

**Reply by wioo** [reply]
Reverse proxy (nginx) has caching disabled and I dont use modsecurity.

**Comment by stefan0xC** [comment]
I've tried to replicate this and the problem is that importing the bitwarden csv of an organization into a user will fail. This also fails on https://vault.bitwarden.com so it's likely a regression.

Until they fixed the bug, you can import the file by renaming `collections` to `folder` (if you don't want to create the collections as folders you'll need to also clear the first entry).

**Reply by wioo** [reply]
Interesting.

my cvs file looks like this (I dont use folders)
```
name,type,login_uri,login_username,login_password
"example.com",login,"https://example.com","name@example.com","some_password"
```

**Reply by stefan0xC** [reply, marked as answer]
Did you create that yourself?  Because by default an export has the following columns:
`folder,favorite,type,name,notes,fields,reprompt,login_uri,login_username,login_password,login_totp`

I've looked it up and the issue was probably introduced in https://github.com/bitwarden/clients/pull/4653 because it does not check if a `folderName` is passed.

So adding an empty `,folder` column (+ a `,` in each line) to the end could be a workaround for this issue.

edit: It seems like it will be fixed by https://github.com/bitwarden/clients/pull/5137

**Reply by wioo** [reply]
Yes, this is my custom format converted from pass (the standard unix password manager).
And you are right. Adding folder column fixed the issue.

Thank you.
